### ISSUE TITLE
Smooth-L1 loss equation differentiable?

### ISSUE THREAD

**Issue by chandlerbing65nm** [open-mmlab/mmdetection#7879]
The equation for `Smooth-L1 loss` is stated as:
![image](https://user-images.githubusercontent.com/62779617/165986821-d7642914-b0c0-446b-a839-23cce4cf632e.png)

To implement this equation in `PyTorch`, we need to use `torch.where()` which is non-differentiable.

```
    diff = torch.abs(pred - target)
    loss = torch.where(diff < beta, 0.5 * diff * diff / beta, diff - 0.5 * beta)
```

Why do we use `torch.where()` for `Smooth-L1 loss` if it is non-differentiable?

**Comment by PeterVennerstrom** [comment]
torch.where() applies a condition to determine the indexes of diff to assign to each loss function.  This indexing does not need to be differentiable since it occurs entirely by heuristics not learned parameters.

**Comment by chandlerbing65nm** [comment]
> torch.where() applies a condition to determine the indexes of diff to assign to each loss function. This indexing does not need to be differentiable since it occurs by heuristics not learned parameters.

@PeterVennerstrom  I don't quite get the explanation. can you please care to elaborate this: 

`since it occurs by heuristics not learned parameters`

**Comment by PeterVennerstrom** [comment]
torch.where() makes selections by index by condition.  There are many forms of indexing used by neural networks which are not differentiable.  Non-differentiable indexing can be thought of as belonging to the architecture definition, not learned parameters.

A simple example would be splitting a tensor in half by channel and sending one half through 'layer_1' and the other half through 'layer_2' before adding the outputs back together.  The indexing selection to split the tensor is not differentiable.  

A better example is dilated convolution vs deformable convolution:

A 3x3 dilated convolution with dilation of 5 uses indexing to select i, j locations set by heuristic (defined by the creator of the architecture not learned).  The indexing selection of i, j locations is not continuous so it is not differentiable and cannot be learned.

A 3x3 deformable convolution is similar to a dilated convolution but the offsets or 'arbitrary dilation' are predicted by the network with learned parameters.  In this case rather than using indexing, the offsets must be sampled from an interpolated feature map to make locations continuous and differentiable.
